### PR TITLE
Use UnionFind for Random.Array.shuffle

### PR DESCRIPTION
**Pull request by owanturist**
Hi Elm Community!

Corresponding changes for `Random.Array.shuffle` is here according #14 .

### CODE REVIEW

File: src/Random/Array.elm [hunk 1]
```diff
@@ -89,27 +90,21 @@ choose arr =
             gen
 
 
-{-| Shuffle the array using the Fisher-Yates algorithm. Takes O(_n_ log _n_)
```
Comment (mgold) [inline src/Random/Array.elm:92]: Keep _array_ please.

File: src/Random/Array.elm [hunk 2]
```diff
@@ -89,27 +90,21 @@ choose arr =
...
-time and O(_n_) additional space.
+{-| Shuffle the list using the Union-Find data structure with path compression algorithm.
+Takes O(_n_ log _n_).
```
Comment (mgold) [inline src/Random/Array.elm]: Can you be more specific as to whether this is time, space, or both?

File: src/Random/Array.elm [hunk 3]
```diff
@@ -89,27 +90,21 @@ choose arr =
...
-        Random.map (Tuple.first >> Array.fromList) (helper ( [], arr ))
+shuffle values =
+    let
+        -- Yes, it takes O(1) time for get the Array's length,
+        -- but let's just keep the value here
+        length =
+            Array.length values
+    in
+    Random.map
+        (Array.fromList << Utils.selectUniqByIndexes values)
```
Comment (mgold) [inline src/Random/Array.elm]: I think `>>` is preferred.

File: tests/Tests/Random/Array.elm [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+                    ( shuffledArray, _ ) =
+                        Random.step (Random.Array.shuffle (Array.fromList initialList)) (Random.initialSeed 0)
+                in
+                shuffledArray
+                    |> Array.toList
+                    |> Expect.equalLists [ 4, 8, 5, 1, 0, 9, 7, 6, 2, 3 ]
+
+        --
+        , Test.fuzzWith { runs = 1 } Fuzz.int "test existing all of the 10k elements" <|
```
Comment (mgold) [inline tests/Tests/Random/Array.elm]: So I get why we're using `runs = 1` with a 10k item array. But could we say, abstract the bodies of each test into functions that take the array size as an argument, and then run a 10 or 100 item test the normal 100 times, and also the 10k test once?

**Review by mgold — CHANGES_REQUESTED** [review]
I hope you don't mind if I'm a bit pickier this time. Where appropriate, please make these changes for `List`, too. Thanks!

File: src/Random/Array.elm (hunk 2 above)
Comment (owanturist) [reply src/Random/Array.elm]: Unfortunately I'm not a big pro of the question. Be honest I've started to lear algorithms and data structures a couple of weeks ago.

The `UnionFind` creates a `Dict` underhood, size of the dict at the end of shuffle will be equal `n - 1`. I assume that is the additional space, isn't it?

File: tests/Tests/Random/Array.elm (hunk 4 above)
Comment (owanturist) [reply tests/Tests/Random/Array.elm]: Sure we could. I just would like keep one test case with 10k for "critical test" so the rest of them I'll make as you proposed.

Comment (mgold) [reply tests/Tests/Random/Array.elm]: Yes, definitely keep the 10k case.

File: src/Random/Array.elm (hunk 2 above)
Comment (mgold) [reply src/Random/Array.elm]: Seems reasonable. So `Takes O(_n_ log _n_) time and space.` then.

**Review by mgold — APPROVED** [review]
Thanks for the fixes!

**Pull request merged**